### PR TITLE
Multi-Scale Intermediate Skips: FPN-style output aggregation

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current architecture is sequential: node features pass through N TransolverBlocks and only the FINAL block's output feeds the prediction head. Earlier blocks capture local/low-frequency features (geometry, wall proximity) while later blocks capture global/high-frequency features (pressure gradients, wake effects).

For pressure prediction, BOTH scales matter: the suction peak is a sharp, local feature (high frequency) while the overall pressure distribution is a global feature (low frequency). A learnable weighted mixture across all blocks allows the model to select the appropriate scale for each output channel.

This is the FPN (Feature Pyramid Network) / U-Net principle — one of the most reliable improvements in image segmentation for tasks requiring multi-scale accuracy. The existing preprocess skip connection (PR #774) demonstrates this architecture is receptive to skip connections.

**Safety:** All skip heads are zero-initialized, so worst case = current baseline (adding zeros).

## Instructions

Add `--multi_scale_skip` flag. Add learnable weighted skip connections from each TransolverBlock to the prediction head.

### Step 1: Add argument

```python
parser.add_argument('--multi_scale_skip', action='store_true',
                    help='FPN-style skip connections from all intermediate TransolverBlocks to output')
```

### Step 2: Define skip projection heads

In the model's `__init__`, after the TransolverBlocks are defined:

```python
if args.multi_scale_skip:
    n_blocks = args.n_layers  # typically 3
    out_dim = 3  # Ux, Uy, p
    self.skip_heads = nn.ModuleList([
        nn.Linear(hidden_dim, out_dim, bias=False) for _ in range(n_blocks)
    ])
    self.skip_scales = nn.ParameterList([
        nn.Parameter(torch.zeros(1)) for _ in range(n_blocks)
    ])
    # Zero-init all skip head weights — starts as identity (adding zeros)
    for head in self.skip_heads:
        nn.init.zeros_(head.weight)
```

### Step 3: Collect intermediate outputs

In the forward pass, collect each TransolverBlock's output:

```python
block_outputs = []
for i, block in enumerate(self.blocks):
    x = block(x, ...)  # existing forward pass
    if args.multi_scale_skip:
        block_outputs.append(x)
```

### Step 4: Add skip predictions to final output

After the final prediction head computes `pred`:

```python
if args.multi_scale_skip:
    for i, (block_out, head, scale) in enumerate(
        zip(block_outputs, self.skip_heads, self.skip_scales)
    ):
        # scale starts at 0.0, gradually learns to contribute
        pred = pred + head(block_out) * torch.sigmoid(scale)
```

Using `sigmoid(scale)` with scale initialized at 0 gives an initial contribution of 0.5 * 0 = 0 (since head weights are zero). As training progresses, both the head weights and scale can grow.

**Alternative (simpler):** Just use the scale directly:
```python
pred = pred + head(block_out) * scale  # scale starts at 0.0
```

Use the simpler version first.

### Step 5: Verification

- At epoch 0: all skip_scales should be ~0, skip contribution should be ~0
- At epoch 50: check if any skip_scales have grown significantly (indicates the model finds intermediate features useful)
- Log `skip_scale_0`, `skip_scale_1`, `skip_scale_2` to W&B

### Training commands

```bash
cd cfd_tandemfoil && python train.py \
  --agent edward --seed 42 \
  --wandb_name "edward/multi-scale-skip-s42" \
  --wandb_group "multi-scale-intermediate-skips" \
  --multi_scale_skip \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73: same flags, --seed 73, --wandb_name "edward/multi-scale-skip-s73"
```

## Baseline

| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.742** | < 11.74 |
| p_oodc | 7.643 | < 7.64 |
| **p_tan** | **27.874** | < 27.87 |
| p_re | 6.419 | < 6.42 |

**Baseline W&B runs:** k5qwvce4 (seed 42), 7oa5xfhi (seed 73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Experiment: Multi-Scale Intermediate Skips (FPN-style output aggregation)

**Implementation:** Added `--multi_scale_skip` flag. For each intermediate TransolverBlock (blocks 0 and 1 with n_layers=3), a zero-initialized `nn.Linear(n_hidden, out_dim=3, bias=False)` projects hidden features to output space. Each has a learnable scalar scale parameter (initialized to 0). The skip contribution is: `pred += head(block_out) * scale`. Both head weights and scales start at 0 → zero contribution at init, gradually learned.

Skip heads are added after `initialize_weights()` to avoid being reset by orthogonal init. Scales are logged per-step to W&B as `skip_scale_0` and `skip_scale_1`.

### Commands used

```bash
# Seed 42 (GPU 0)
python train.py --agent edward --seed 42 --wandb_name "edward/multi-scale-skip-s42" \
  --wandb_group "multi-scale-intermediate-skips" --multi_scale_skip \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73: same with CUDA_VISIBLE_DEVICES=1, --seed 73, --wandb_name "edward/multi-scale-skip-s73"
```

### Metrics vs Baseline

Both runs completed 143 epochs (180-min wall-clock limit).

| Metric | s42 | s73 | **2-seed avg** | Baseline | Delta |
|--------|-----|-----|------------|---------|-------|
| **p_in** | 12.6 | 12.1 | **12.35** | 11.742 | +0.61 ❌ |
| **p_oodc** | 8.1 | 7.7 | **7.90** | 7.643 | +0.26 ❌ |
| **p_tan** | 28.8 | 29.4 | **29.10** | 27.874 | +1.23 ❌ |
| **p_re** | 6.6 | 6.4 | **6.50** | 6.419 | +0.08 ❌ |

W&B run IDs: [edward/multi-scale-skip-s42](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/3imchppo) | [edward/multi-scale-skip-s73](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/i2exxwpa)

Peak memory: 47.5 GB (s42), 46.9 GB (s73)

### What happened

**Negative result.** All metrics are worse than baseline. p_tan is the worst affected (+1.23).

Analysis:

1. **The skip scales and zero-init weights mean the contribution remains small throughout training.** With Lion optimizer (adaptive, no momentum-based acceleration), the zero-init weights + zero scales create a very flat loss landscape around zero — the gradient signal to activate the skip paths must overcome both the near-zero head weights AND the near-zero scale simultaneously. This creates a "chicken and egg" problem: the head can't learn useful projections because the scale suppresses gradients, and the scale can't grow because the head produces near-zero output.

2. **Architecture already has skip connections.** The existing `out_skip + skip_gate` path from `fx_pre` (preprocess output) to the final prediction already provides a residual skip connection across all blocks. Adding more intermediate skips may not provide additional signal the model can't already learn through this existing path.

3. **Intermediate blocks produce latent representations, not prediction-aligned features.** The non-last TransolverBlocks don't have output heads — they produce hidden features optimized for the NEXT block's input, not for direct prediction. The skip heads must learn to extract prediction-relevant information from representations that aren't shaped for that purpose.

4. **3 blocks is a shallow network.** The FPN/U-Net benefit is strongest in deep networks (e.g. ResNet-50 with 50+ layers) where early and late features are very different. With only 3 TransolverBlocks, the feature hierarchy is too shallow for multi-scale aggregation to add value.

### Suggested follow-ups

- **Warmup the scale instead of zero-init:** Start scale at 0.1 (or even 0.5) to give the heads gradient signal from the start, rather than requiring both head weights and scale to co-evolve from zero.
- **Use sigmoid(scale) instead of bare scale** (as the PR also suggested): sigmoid(0)=0.5, giving immediate 50% contribution.
- **Skip from preprocess output at different positions** in the attention computation rather than between blocks — e.g., skip from pre-attention vs post-attention within the same block.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2313 because ADVISOR: Closing — all metrics worse (p_in +5.2%, p_tan +4.4%).

**Your analysis is correct on all points:**
1. With only 3 TransolverBlocks, the network is too shallow for FPN-style multi-scale aggregation to help. The feature hierarchy isn't deep enough.
2. The zero-init chicken-and-egg problem with Lion optimizer is a real issue — both head weights and scale must co-evolve from zero.
3. The existing preprocess skip already provides cross-block residual connectivity.

Added to exhausted list: multi-scale intermediate skips are ineffective with 3 blocks. New assignment coming..
